src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala

```diff
@@ -0,0 +1,239 @@
+      case None ⇒
+        Props(new CurrentEventsByTagPublisher(tag, fromOffset, maxBufSize, session, preparedSelect))
+    }
+  }
+
+  private[query] case object Continue
+
+  private[query] final case class TaggedEventEnvelope(
+    offset: UUID,
+    persistenceId: String,
+    sequenceNr: Long,
+    event: Any)
+    extends DeadLetterSuppression with NoSerializationVerificationNeeded
+
+  private[query] case class ReplayDone(count: Int)
+    extends DeadLetterSuppression
+  private[query] final case class ReplayFailed(cause: Throwable)
+    extends DeadLetterSuppression with NoSerializationVerificationNeeded
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[query] abstract class AbstractEventsByTagPublisher(
+  val tag: String, val fromOffset: UUID,
+  val maxBufSize: Int, val session: Session, val preparedSelect: PreparedStatement)
+  extends ActorPublisher[EventEnvelope] with DeliveryBuffer[EventEnvelope] with ActorLogging {
+  import EventsByTagPublisher._
+
+  var currTimeBucket: String = CassandraJournal.timeBucket(UUIDs.unixTimestamp(fromOffset))
+  var currOffset: UUID = fromOffset
+
+  def nextTimeBucket(): Unit = {
+    val nextDay = LocalDate.parse(currTimeBucket, CassandraJournal.timeBucketFormatter).plusDays(1)
+    currTimeBucket = nextDay.format(CassandraJournal.timeBucketFormatter)
+  }
+
+  def today: LocalDate = LocalDate.now(ZoneOffset.UTC)
+
+  def bucketDate: LocalDate = LocalDate.parse(currTimeBucket, CassandraJournal.timeBucketFormatter)
+
+  def isTimeBucketBeforeToday(): Boolean =
+    bucketDate.isBefore(today)
+
+  def isTimeBucketTodayOrLater(): Boolean = {
+    val tday = today
+    val bucket = bucketDate
+    bucket.isEqual(tday) || bucket.isAfter(tday)
+  }
+
+  // exceptions from Fetcher
+  override val supervisorStrategy = OneForOneStrategy() {
+    case e =>
+      self ! ReplayFailed(e)
+      SupervisorStrategy.Stop
+  }
+
+  def receive = init
+
+  def init: Receive = {
+    case _: Request ⇒ receiveInitialRequest()
+    case Continue   ⇒ // skip, wait for first Request
+    case Cancel     ⇒ context.stop(self)
+  }
+
+  def receiveInitialRequest(): Unit
+
+  def idle: Receive = {
+    case Continue ⇒
+      if (timeForReplay)
+        replay()
+
+    case _: Request ⇒
+      receiveIdleRequest()
+
+    case Cancel ⇒
+      context.stop(self)
+  }
+
+  def receiveIdleRequest(): Unit
+
+  def timeForReplay: Boolean =
+    (buf.isEmpty || buf.size <= maxBufSize / 2)
+
+  def replay(): Unit = {
+    val limit = maxBufSize - buf.size
+    log.debug("query for tag [{}] from [{}] [{}] limit [{}]", tag, currTimeBucket, currOffset, limit)
+    context.actorOf(EventsByTagFetcher.props(tag, currTimeBucket, currOffset, limit, self, session, preparedSelect))
+    context.become(replaying(limit))
+  }
+
+  def replaying(limit: Int): Receive = {
+    case TaggedEventEnvelope(offs, pid, seqNr, evt) ⇒
+      // FIXME handle seqNr in wrong order, due to Materialized View async eventually consistency
```

Comment:
@zapletal-martin do you have any thoughts about this FIXME? I guess that I have to keep a `Map` of sequence number per persistenceId.